Neos.Media/Classes/Domain/Service/AssetVariantGenerator.php

```diff
@@ -0,0 +1,112 @@
+<?php
+namespace Neos\Media\Domain\Service;
+
+/*
+ * This file is part of the Neos.Media package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\Configuration\Exception\InvalidConfigurationException;
+use Neos\Flow\ResourceManagement\Exception;
+use Neos\Media\Domain\Model\Adjustment\ImageAdjustmentInterface;
+use Neos\Media\Domain\Model\AssetInterface;
+use Neos\Media\Domain\Model\Image;
+use Neos\Media\Domain\Model\ImageVariant;
+use Neos\Media\Domain\ValueObject\Configuration;
+use Neos\Media\Exception\AssetVariantGeneratorException;
+use Neos\Media\Exception\ImageFileException;
+use Neos\Utility\ObjectAccess;
+
+/**
+ * @Flow\Scope("singleton")
+ */
+class AssetVariantGenerator
+{
+    /**
+     * @Flow\Inject
+     * @var AssetService
+     */
+    protected $assetService;
+
+    /**
+     * @param AssetInterface $asset
+     * @return ImageVariant[] The created variants (if any), with the preset identifier as array key
+     * @throws AssetVariantGeneratorException
+     * @throws Exception
+     * @throws ImageFileException
+     * @throws InvalidConfigurationException
+     */
+    public function createVariants(AssetInterface $asset): array
+    {
+        if (!$asset instanceof Image) {
+            return [];
+        }
+
+        $createdVariants = [];
+        foreach ($this->assetService->getImageVariantPresets() as $presetIdentifier => $preset) {
+            foreach ($preset->variants() as $variantConfiguration) {
+                $createdVariants[$presetIdentifier] = $this->createVariant($asset, $presetIdentifier, $variantConfiguration);
+                $asset->addVariant($createdVariants[$presetIdentifier]);
+            }
+        }
+
+        return $createdVariants;
+    }
+
+    /**
+     * @param Image $imageAsset
+     * @param string $presetIdentifier
+     * @param Configuration\Variant $variantConfiguration
+     * @return ImageVariant
+     * @throws AssetVariantGeneratorException
+     * @throws Exception
+     * @throws ImageFileException
+     * @throws InvalidConfigurationException
+     * @throws \Exception
+     */
+    protected function createVariant(Image $imageAsset, string $presetIdentifier, Configuration\Variant $variantConfiguration): ImageVariant
+    {
+        $adjustments = [];
+        foreach ($variantConfiguration->adjustments() as $adjustmentConfiguration) {
+            assert($adjustmentConfiguration instanceof Configuration\Adjustment);
```

Comment:
First and only `assert()` in Neos code. Remove it or configure assert according to Neos contexts?

Reply:
The assert is meant to help IDEs with type hinting, instead of the more ugly /** @var …**/ inline comments. Is it too confusing?

Reply:
No, not confusing - interesting. I never saw it before. Just read through http://php.net/manual/en/function.assert.php and I think if we use it, we should configure the behavior according to our Development / Production contexts. WDYT?

Reply:
That's also interesting, I didn't know that it can be configured! ;-)

So far I only use assert for cases where I am already sure what the type will be, so I don't expect them to fail ever. I mostly use them in loops and the like, especially when I don't have influence on the code defining the array I'm looping over. So, if it was just me, I'd be fine with just using the default configuration and let other developers decide if they want to tweak it through php.ini.

Reply:
FYI: https://twitter.com/localheinz/status/993000253270917120